### ISSUE TITLE
Limitations and confusions with parameter AuthenticationMethod in resources SPWebApplication and SPWebApplicationExtension

### ISSUE THREAD

**Issue by Yvand** [PowerShell/SharePointDsc#538]
SharePointDsc 1.6 introduces possibility to specify federated authentication in SPWebApplication resource, which is great, but current implementation has multiple issues:
- Parameter AuthenticationMethod accepts "NTLM", "Kerberos" and "Claims", but all (NTLM, Kerberos, FBA or Federated) are actually claims authentication, so possible values should be "NTLM", "Kerberos" and "Federated".
- FBA option is missing.
- Parameter AutheticationMethod does not allow to combine multiple authentication modes.

The same mistakes are being made in [upcoming resource SPWebApplicationExtension](https://github.com/PowerShell/SharePointDsc/pull/510)

**Comment by mrpullen** [comment]
Should we create a resource for "**SPWebApplicationAuthenticationMethod**" so each authentication method can be captured for each web application as a specific resource? From my perspective each "setting" should be an independent resource, instead of trying to control multiple settings in a single resource..

thoughts?

we could update the **SPWebApplicationAuthenticationMethod** as you suggest, and enable you to submit a named Provider in the case of Federated, then provide a separate resource to manage other authentication methods.

**Comment by BrianFarnhill** [comment]
@mrpullen That's one option, or we could capture it as a complex type property within the web app (Have a look at how we do things like [SPServiceAppSecurity](https://github.com/PowerShell/SharePointDsc/wiki/SPServiceAppSecurity) to see what I mean). This would keep it in the one resource, but we need to look at the number of properties and complexity of managing the authentication providers I think. If they are gonna be too big then yes a seperate resoruce is warranted and we refer to the provider by name as a property on SPWebApplication.

**Comment by mrpullen** [comment]
So it would be an array of those objects? How would that work and be "backward compatible?". 

I guess the object would look like --- 
```
@{
   AuthenticationMethod = "" //NTLM, Kerberos, FBA, Federated
   AuthenticationProvider = "" //Name of TrustedIdentityTokenIssuer if Federated.
   MembershipProvider = "" //Name of Membership Provider if using FBA
   RoleManager = "" //Name of the Role Manager if using FBA
   SQLServer = "" //Name of the SQL Server the Membership Provider / Role Manager sqldb are saved.
   SQLDatabase = "" // Name of the SQL Server Database.
}
```
maybe some of other properties from [Configure Forms Based Authentication on SharePoint 2013](https://sharepointumar.wordpress.com/2013/08/01/configure-forms-based-authentication-for-sharepoint-2013/)
Then we could have separate resources to actually initialize the Membership and RoleManager. Not sure if this belongs in the SharePointDSC pack, perhaps another resource for ASP.NET App Settings, then update the SPWebApplication and SPWebApplicationExtension with code to make the necessary inserts into the web.config??

**Comment by BrianFarnhill** [comment]
In terms of backwards compatability - put simply it wont. The second we make a change to this we are breaking the choices/options there now so this will be a change we put in a our v2.0 release branch as a breaking change, so I wouldn't worry about that. 

In terms of the FBA stuff and what goes in to SharePoint vs what hits the web.config file - I would focus on making sure that the SharePoint settings are in our resources, where as the web.config changes hit another resource (I'm pretty sure there is something about xml file manipulation in the works in another open source resource for this, I'll follow that up).

**Comment by mrpullen** [comment]
Ok. So if we leave off the web.config stuff the object would be

This would be a MSFT_SPAuthentication object?
```
@{
   AuthenticationMethod = "" //NTLM, Kerberos, FBA, Federated
   AuthenticationProvider = "" //Name of TrustedIdentityTokenIssuer if Federated.
   MembershipProvider = "" //Name of Membership Provider if using FBA
   RoleManager = "" //Name of the Role Manager if using FBA
 }
```

**Comment by BrianFarnhill** [comment]
At a first glance yep that seems how I would do it

**Comment by ykuijs** [comment]
Has this been fully implemented? If not, what functionality is still needed?

**Comment by NikCharlebois** [comment]
I like the idea of using a Custom Object to address this. However, how will the SPAuthenticationProvider be mapped to a security zone? Would that be a property of the SPAuthentication object?

**Comment by ykuijs** [comment]
This is the only issue that is open for a v2.0 release. After having a discussion with @NikCharlebois, we came to the conclusion that we need a new resource. When using a complex type, we have no way of preventing adding two (possible conflicting) configurations for a web application zone.

Therefore we decided to create a new resource to configure the authentication models for all web applications and extensions. This resource has two Key parameters: Url and Zone, and one complex type array parameter with the authentication methods.

**Comment by BrianFarnhill** [comment]
@ykuijs If it's a new resource, is this still a breaking change from the original resource?

**Comment by ykuijs** [comment]
It isn't just a new resource, it is also a removing the current properties on SPWebApplication and SPWebApplicationExtension which are breaking changes

**Comment by ykuijs** [comment]
Fix included in PR #690

**Comment by ykuijs** [comment]
PR merged, closing issue
